crates/fixt/src/rng.rs

```diff
@@ -0,0 +1,58 @@
+//! Seedable random number generator to be used in all fixturator randomness
+//!
+//! In tests, when an unpredictable value causes a test failure, it's important to
+//! be able to re-run the test with the same values. This module provides a RNG
+//! whose seed will be set automatically and printed to stdout before each test run.
+//! To use a previous seed, just set the FIXT_SEED environment variable to the value
+//! of a previous run'
+
+use parking_lot::Mutex;
+use rand::SeedableRng;
+use rand::{rngs::StdRng, RngCore};
+use std::sync::Arc;
+
+lazy_static::lazy_static! {
+    /// The singleton global RNG for test randomness
+    pub static ref FIXT_RNG: FixtRng = {
+        let seed: u64 = match std::env::var("FIXT_SEED") {
+            Ok(seed_str) => {
+                seed_str.parse().expect("Expected integer for FIXT_SEED")
+            }
+            Err(std::env::VarError::NotPresent) => { rand::random() },
+            Err(std::env::VarError::NotUnicode(v)) => { panic!("Invalid FIXT_SEED value: {:?}", v) },
+        };
+        println!("Fixturator seed: {}", seed);
+        FixtRng(Arc::new(
+            Mutex::new(StdRng::seed_from_u64(seed))
+        ))
+    };
+}
+
+/// A seedable RNG which uses an Arc and a Mutex to allow easy cloneability and thread safety.
+/// A singleton global instance is created in this module. See module-level docs for more info.
+#[derive(Clone)]
+pub struct FixtRng(Arc<Mutex<StdRng>>);
```

Comment:
I'm a little worried about the deadlock potential here. I don't think you need to share the same `StdRng` across all the fixtures. You just need to create them from the same seed.

Reply:
@freesig if we just recreate rngs with the same seed and throw them away, won't we get the same value every time? I think we have to share the rng and let it be mutated over time, that's exactly how it generates randomness.

There's no way to deadlock here because there's no way to explicitly get a lock on the underlying mutex, since the internals are not public. This is why I implemented Rng for the struct rather than having consumers do unlocking themselves.

The fact that the tests pass I think proves that this is no problem. We have deeply nested fixturators that are all getting locks via the Rng interface.

Reply:
Ah yeh good point. I was thinking the fixturators consumed the lock but I see they mostly take a reference to it